In [32]:
import pandas as pd
import numpy as np
import seaborn as sns

#import statsmodels formula
import statsmodels.formula.api as smf
import statsmodels.api as sm


# Økonometri A  

## Problem Set 6  

### Hedonic price regressions  

In problem set 6, we estimate a regression model relating house prices to house characteristics. This model is an example of the so-called hedonic price regression which is widely used in economics.

A hedonic regression for house prices usually includes house characteristics and community attributes as explanatory variables. In this case, the model's coefficients may be interpreted as the implicit price of each characteristic. Hedonic price models can be useful for estimating the price of characteristics for which there are no markets. For example, we do not observe a price for clean air, but we may be able to estimate the (implicit) price effect of clean air on house prices.

The data used in problem set 6 contains a random sample of apartment sales in Copenhagen in 2005. We will focus on apartments sold in the four neighborhoods of Copenhagen K, N, V, and Ø. For each apartment, we observe the sales price and a range of apartment characteristics which are all specific to the year of 2005. We consider these data as a cross-section, exploiting variation in prices and characteristics across apartments to estimate the parameters of the hedonic price regression.

The STATA file `PS6.dta` includes the following variables for a total of 988 apartment sales in 2005:

- Sales price in 2005-DKK (**price**)
- Apartment size in square meters (**m2**)
- Number of rooms (**rooms**)
- Number of toilets (**toilets**)
- Floor location of apartment (**floor**)
- Apartment location in Copenhagen (**location**)
- Number of apartment units in the building (**building_units**)
- Building age (**age**)


---

### Group work  

Discuss the following questions in groups:

#### Question 1.

Consider the simple hedonic model:

$$
\log(price_i) = \beta_0 + \beta_1 m2_i + \beta_3 rooms_i + \beta_4 toilets_i + u_i \tag{1}
$$

 Discuss if it is reasonable to assume that assumptions MLR.1–MLR.5 are satisfied for model (1).



**Your answer:**

Ja, det kan vi godt.


#### Question 2.
 The hedonic model can be extended with dummy variables to investigate if there are level differences between apartments in different neighbourhoods of Copenhagen. Note that the variable **location** takes on four categories:
 _KBH K_, _KBH N, KBH O_ and _KBH V_

 The extended model with dummy variables could look like this:
 
 \begin{align}
 		\log(price_i)=\, &\beta_0 +\delta_1 KbhN_i + \delta_2 KbhO_i + \delta_3 KbhV_i \\
 			      &+\beta_1 m2_i + \beta_2 rooms_i + \beta_3 toilets_i 
 			      + \epsilon_i 
 	\end{align}\tag{2}

1. Explain what a dummy variable is.

2. Why don't we include dummies for _Kbh K_ model (2)? 

3. In terms of the model parameters, what is the intercept for apartments located in Kbh V?

**Your answer:**

1. I dette tilfælde er det vores deltaer. De fungerer som indikator funktion til at få der rette ting med. 

2. Fordi kbh K er referencegruppen. 

3. $\beta_0 + \delta_3$

### Question 3. 
The hedonic model can further be extended with interaction terms to see if the model parameters differ across locations in Copenhagen.

\begin{align}
 		\log(price_i)=\, &\beta_0 +\delta_1 KbhN_i + \delta_2 KbhO_i + \delta_3 KbhV_i \\
 			      &+\beta_1 m2_i  +\delta_4 KbhN_i\cdot m2_i  + \delta_5 KbhO_i\cdot m2_i  + \delta_6 KbhV_i\cdot m2_i  \\
                  &+\beta_2 rooms_i +\delta_7 KbhN_i\cdot rooms_i  + \delta_8 KbhO_i\cdot rooms_i + \delta_9 KbhV_i\cdot rooms_i \\
 		          &+ \beta_3 toilets_i +\delta_{10} KbhN_i\cdot toilets_i  + \delta_{11} KbhO_i\cdot toilets_i + \delta_{12} KbhV_i\cdot toilets_i \\
 			      & + \epsilon_i 
 	\end{align}\tag{3}

1. Which coefficients in model (3) describe the interaction terms?

2. In terms of the model parameters, what is the expected log(price) for a  $ 75 m^2$ apartment in _Kbh V_ with three rooms and one toilet?

3. What is the expected log(price) for an identical apartment in _Kbh K_?

1. $\delta_4$ til $\delta_{12}$

2. $E(\log(price))_i=\beta_0+\beta_1\cdot75+\beta_2\cdot3+\beta_3+\delta_3+\delta_6\cdot75+\delta_9\cdot3+\delta_{12}$

3. Samme som foroven


#### Question 4.
 How can you test for level differences in apartment prices across Copenhagen? How can you test if the (implicit) price of an additional square meter is different across locations in Copenhagen? Formulate the null and alternative hypotheses (be precise!)


**Your answer:**

Vigtigt at det er samme slags bolig på tværs af stederne, altså holder kvadratmeter, toilets og rooms fast.

For at teste vil vi bruge model 3, og teste om vores forskellige steder har inflydsele på prise.

$H_0$: $\delta_1=...=\delta_{12}=0$

$H_a$: er at blot en af dem er forskellig fra 0.


---

### Python exercises


#### Task 0: Warm-up

When using Statsmodels to perform hypothesis testing in this problem set, it will be useful for you to know a little about two very useful Python features, namely **f-strings** and **list comprehension**.



##### f-strings 
f-strings lets you plug Python variables directly into your strings. Consider the example below:
```py
name = 'Daniel'
age = 30 + 1
greeting = f'My name is {name}, great to meet you! I am {age} years old'

print(greeting)
```

```txt
>> My name is Daniel, great to meet you! I am 31 years old
```

So by simply adding an 'f' in front of your strings, you get the superpower of being able to include the contents of variables directly in your strings.




##### List comprehension
List comprehension is another useful tool that allows you quickly to generate transformations of existing lists without needing a for loop:
```py
numbers = [1, 2, 3, 4]
numbers2 = [2 * num for num in numbers] # <- list comprehension

print(numbers)
```

```txt
>> [2, 4, 6, 8]
```
List comprehension can also be used to work with strings. And you can loop over multiple lists in the same list comprehension statement. Consider this example:

```py
letters = ['x', 'y', 'z']
numbers = [1, 2, 3, 4]

variables = [f'{let}_{num}' for let in letters for num in numbers]

print(variables)
```

```txt
>> ['x_1', 'x_2', 'x_3', 'x_4', 'y_1', 'y_2', 'y_3', 'y_4', 'z_1', 'z_2', 'z_3', 'z_4']

**Task:** Use your knowledge of list comprehension and f-strings to generate this output from the two lists in the code cell below:

```py
['C(location)[T.KBH N]:m2',
 'C(location)[T.KBH N]:rooms',
 'C(location)[T.KBH N]:toilets',
 'C(location)[T.KBH O]:m2',
 'C(location)[T.KBH O]:rooms',
 'C(location)[T.KBH O]:toilets',
 'C(location)[T.KBH V]:m2',
 'C(location)[T.KBH V]:rooms',
 'C(location)[T.KBH V]:toilets']
```


In [9]:
locs = ['KBH N', 'KBH O', 'KBH V']
vars = ['m2', 'rooms', 'toilets']
loc_vars = [f'C(location)[T.{loc}]:{var}' for loc in locs for var in vars]    # Your list comprehension code here
print(loc_vars)


['C(location)[T.KBH N]:m2', 'C(location)[T.KBH N]:rooms', 'C(location)[T.KBH N]:toilets', 'C(location)[T.KBH O]:m2', 'C(location)[T.KBH O]:rooms', 'C(location)[T.KBH O]:toilets', 'C(location)[T.KBH V]:m2', 'C(location)[T.KBH V]:rooms', 'C(location)[T.KBH V]:toilets']


#### Task 1.
**Load the data set into pandas** and provide a descriptive analysis of sales prices and apartment sizes across different locations in Copenhagen.

_Hint 1:_ Remember from Problem Set 1 that we can compute grouped summary statistics by using the `.groupby()` method in a DataFrame.

_Hint 2:_ If you don't want all the summary statistics, but just the mean, you can use `.mean()` instead of `.describe()`. This is especially useful when grouping on some category, as the resulting table has a tendency of becoming very big otherwise. Similar useful methods are `.std()`, `.count()`, `.max()`, `.min()` and `.median()`

_Hint 3:_ You can use `df['location'].value_counts()` to see the distribution of the observations across the four locations in the dataset.


**Your code:**

In [22]:
df = pd.read_stata('PS6.dta')

df.groupby('location').describe()


price                                                              \
          count          mean           std       min        25%        50%   
location                                                                      
KBH K     223.0  2.625185e+06  1.064166e+06  750000.0  1840000.0  2495000.0   
KBH N     229.0  1.672342e+06  6.111055e+05  644000.0  1245000.0  1560000.0   
KBH O     398.0  2.404902e+06  9.951280e+05  563772.0  1695000.0  2200000.0   
KBH V     138.0  2.252893e+06  9.474386e+05  722500.0  1570000.0  2040000.0   

                                   m2             ... building_units         \
                75%        max  count       mean  ...            75%    max   
location                                          ...                         
KBH K     3167500.0  5995000.0  223.0  90.448430  ...          29.50  162.0   
KBH N     1950000.0  4345000.0  229.0  66.445415  ...          58.00  267.0   
KBH O     2950000.0  6000000.0  398.0  89.655779  ...          60.75  249.0   
KBH V     2895517.5  5050000.0  138.0  85.920290  ...          45.50  163.0   

            age                                                          \
          count         mean        std     min     25%     50%     75%   
location                                                                  
KBH K     223.0  1901.067265  46.418096  1850.0  1859.5  1888.0  1938.0   
KBH N     229.0  1923.519651  33.818735  1859.0  1899.0  1908.0  1947.0   
KBH O     398.0  1936.208543  41.024097  1869.0  1903.0  1926.0  1975.5   
KBH V     138.0  1915.028986  34.055545  1859.0  1895.0  1904.0  1938.0   

                  
             max  
location          
KBH K     2007.0  
KBH N     1991.0  
KBH O     2007.0  
KBH V     2008.0  

[4 rows x 56 columns]

**Your answer:**

Ser at det gennemsnitligt er dyrest at bo på Kbh k og billigst på nørrebro. Samt at kvadratmeter er mindst på nørrebro.


#### Task 2.
**Assume model (1) satisfies MLR.1–MLR.5.** Estimate model (1) by OLS and comment on the parameter estimates. How much of the variation in $\log(price)$ can the regression model explain? Is the sign of $\hat{\beta}_3$ surprising?


**Your code:**

In [39]:
df['logp'] = np.log(df.price)

model1 = smf.ols('logp ~ m2 + rooms + toilets', data=df).fit()
print(model1.summary())

                            OLS Regression Results                            
Dep. Variable:                   logp   R-squared:                       0.562
Model:                            OLS   Adj. R-squared:                  0.561
Method:                 Least Squares   F-statistic:                     421.1
Date:                Tue, 01 Oct 2024   Prob (F-statistic):          6.51e-176
Time:                        16:16:19   Log-Likelihood:                -148.92
No. Observations:                 988   AIC:                             305.8
Df Residuals:                     984   BIC:                             325.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.9043      0.042    328.925      0.0

**Your answer:**

$R^2=0.562$

Ser at det er dyrere jo flere kvadratmereter og værelser du har. Samt billigere jo flere toiletter du har, hvilket virker mærkeligt, men havde toilettet været et andet værelse havde det klart været dyrere. 


#### Task 3.
- Estimate model (2)

- Interpret the regression results. What is the estimated price differences across neighbourhoods in percentages?

Make sure you understand both hints below before you solve the task.





[ _Hint:_ Assuming your DataFrame is named `df`, this code may be of help when constructing the dummy variables:
 ```py
 df['KbhN'] = df.location == "KBH N"
 ```
 This creates a new column called `KbhN` that is filled with True or False values depending on whether each observation satisfies the condition. When including this variable in a statsmodels regression, it will automatically be interpreted as a dummy (True is 1, False is 0). 
 
 However, if you want to, you can add this line of code to convert the boolean array to dummies using this code:
 ```py
 df['KbhN'] = df['KbhN'].astype('int')
 ```
]


 
 [_Hint 2:_ Actually you don't have to generate the dummy variables manually.
 
 If you are using formulas to specify your model in statsmodels (as we learned in Problem Set 5), you can skip the process described in the former hint entirely and simply add `C(location)` to your formula string to automatically add dummies based on the location categories to your regression model. Statsmodels also automatically leaves out one category to avoid the dummy trap.
 
  You can read more about this feature at https://www.statsmodels.org/stable/example_formulas.html#categorical-variables
  ]


**Your code:**

In [40]:
model2 = smf.ols('logp ~ C(location) + m2 + rooms + toilets', data=df).fit()
print(model2.summary())


                            OLS Regression Results                            
Dep. Variable:                   logp   R-squared:                       0.597
Model:                            OLS   Adj. R-squared:                  0.595
Method:                 Least Squares   F-statistic:                     242.4
Date:                Tue, 01 Oct 2024   Prob (F-statistic):          8.58e-190
Time:                        16:23:36   Log-Likelihood:                -107.69
No. Observations:                 988   AIC:                             229.4
Df Residuals:                     981   BIC:                             263.6
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept               14.0421 

**Your answer:**

Ser at det er 23% billigere at bo i Kbh N end Kbh K.

Ser at det er 8% billigere at bo i Kbh O end Kbh K.

Ser at det er 13% billigere at bo i Kbh V end Kbh K.

### Task 4

Use an $F$-test if there are level differences in apartment prices across neighbourhoods of Copenhagen. Be precise in formulating your null and alternative hypothesis.

- Why are we using an $F$-test?

- What assumptions are necessary for the validity of the test?

- Based on your test results, do you prefer model (1) or model (2)?



_Hint:_ You can calculate the F-test by hand or use the built-in `.f_test()` function of your statsmodels OLS results object. For example, if you want to test if the coefficiants for $m2$ and $rooms$ are both equal to zero, you would use the code

```py
ftest = results.f_test(['m2', 'rooms'])

print(ftest)
```

In [45]:
model2 = smf.ols('logp ~ C(location) + m2 + rooms + toilets', data=df).fit()
ftest = model2.f_test(['C(location)[T.KBH N]','C(location)[T.KBH O]','C(location)[T.KBH V]'])
print(ftest)

<F test: F=28.46532700225673, p=1.1870777814743184e-17, df_denom=981, df_num=3>


**Your answer:**

$H_0: C(location)=0$

* Bruger F-test da vi gerne vil teste flere hypoteser på samme tid. 

* De der MLR

* Model 2, da vi netop har vist der er forskel på prisen afhængigt af hvor du bor. Hvilket model 1 jo ikke tog højde for. 


#### Task 5.
**Interaction terms for apartment size and location.** 
1. Estimate the full model (3) with all the interaction terms. Interpret the regression results. Are all the estimated coefficiants individually significant?

3. Test whether there are interaction effects across locations in Copenhagen (test if all the interaction terms are jointly 0)

4. Test specifically whether the effect of the number of rooms (`rooms`) differs across locations.

5. Test specifically whether the price effect of apartment size (`m2`) differs across locations.

6. If you want to, try to estimate a new specification based on your test insights to better explain the data.

_Hint:_ If you use formulas to specify your regression model in statsmodels, you can interact two terms in the formula by using the `*` operator instead of `+`. 

For example, if I wanted to interact `m2` with `rooms`, I could use the code:
```py
results = smf.ols('logprice ~ toilets + m2 * rooms', data = df).fit()
```
You can also interact a variable with multiple variables by grouping them in parantheses. For example:
```py
results = smf.ols('logprice ~ m2*(rooms + toilets)', data = df).fit()
```

Note that when interacting two variables, statsmodels automatically adds the two variables individually (that is, un-interacted) to the model specification too.


_Hint:_ Use the output from the warmup exercise (on f-strings and list comprehension) to conduct the first F-test. 

**5.1 code:**

**5.1 answer:**

**5.2 code:**

**5.2 answer:**

**5.3 code:**

**5.3 answer:**

**5.4 code:**

**5.4 answer:**


#### Task 6.
**Quadratic model for apartment size.** 
Model (1) assumes that apartment prices depend linearly on apartment size, but this may be a restrictive assumption. 

You are therefore asked to estimate a new model including a quadratic term of $m2$. Moreover, the model should allow the effect of the linear and quadratic **m2** terms to be different across locations, while making the simplifying assumption that **rooms** and **toilets** have the same effect on sales prices across locations. 

- Generate the variable for squared m2 and estimate the new model
- Test the new model against a restricted model where all slope parameters are the same across locations in Copenhagen. 
- Which model do you prefer in this case?
- What is the expected effect of increasing $m2$ by one unit on prices?

_Hint:_ It might be helpful to scale the squared term by a factor of e.g. 1000 (how will this affect your estimates?)

**Your code:**

**Your answer:**